File: src/main/java/no/nav/syfo/api/selvbetjening/controller/BrukerTilgangController.java
package no.nav.syfo.api.selvbetjening.controller;

import lombok.extern.slf4j.Slf4j;
import no.nav.security.oidc.context.OIDCRequestContextHolder;
import no.nav.security.spring.oidc.validation.api.ProtectedWithClaims;
import no.nav.syfo.api.selvbetjening.domain.RSTilgang;
import no.nav.syfo.brukertilgang.BrukertilgangConsumer;
import no.nav.syfo.metric.Metrikk;
import no.nav.syfo.service.BrukertilgangService;
import no.nav.syfo.service.PersonService;
import org.springframework.util.MultiValueMap;
import org.springframework.util.StringUtils;
import org.springframework.web.bind.annotation.*;

import javax.inject.Inject;
import javax.ws.rs.ForbiddenException;

import static no.nav.syfo.oidc.OIDCIssuer.EKSTERN;
import static no.nav.syfo.oidc.OIDCUtil.getSubjectEksternMedThrows;
import static no.nav.syfo.util.HeaderUtil.NAV_PERSONIDENT;

@Slf4j
@RestController
@ProtectedWithClaims(issuer = EKSTERN)
@RequestMapping(value = "/api/tilgang")
public class BrukerTilgangController {

    final static String IKKE_TILGANG_GRUNN_DISKRESJONSMERKET = "kode6-7";

    private final OIDCRequestContextHolder contextHolder;
    private final BrukertilgangConsumer brukertilgangConsumer;
    private final BrukertilgangService brukertilgangService;
    private final PersonService personService;
    private final Metrikk metrikk;

    @Inject
    public BrukerTilgangController(
            OIDCRequestContextHolder contextHolder,
            PersonService personService,
            BrukertilgangConsumer brukertilgangConsumer,
            BrukertilgangService brukertilgangService,
            Metrikk metrikk
    ) {
        this.contextHolder = contextHolder;
        this.brukertilgangConsumer = brukertilgangConsumer;
        this.brukertilgangService = brukertilgangService;
        this.personService = personService;
        this.metrikk = metrikk;
    }

    @GetMapping
    public RSTilgang harTilgang(@RequestParam(value = "fnr", required = false) String oppslaattFnr) {
        String innloggetIdent = getSubjectEksternMedThrows(contextHolder);
        String oppslaattIdent = StringUtils.isEmpty(oppslaattFnr) ? innloggetIdent : oppslaattFnr;

        if (!brukertilgangService.tilgangTilOppslattIdent(innloggetIdent, oppslaattIdent)) {
            log.error("Ikke tilgang: Bruker spør om noen andre enn seg selv eller egne ansatte");
            throw new ForbiddenException();
        }

        metrikk.tellHendelse("sjekk_brukertilgang");

        if (personService.erDiskresjonsmerket(oppslaattIdent)) {
            return new RSTilgang()
                    .harTilgang(false)
                    .ikkeTilgangGrunn(IKKE_TILGANG_GRUNN_DISKRESJONSMERKET);
        }
        return new RSTilgang().harTilgang(true);
    }

    @GetMapping(path = "/ansatt")
    public Boolean accessToAnsatt(@RequestHeader MultiValueMap<String, String> headers) {
        headers.forEach((key, value) -> log.info("JTRACE key {}, value {}", key, value));
        String oppslaattIdent = headers.getFirst(NAV_PERSONIDENT.toLowerCase());

        if (StringUtils.isEmpty(oppslaattIdent)) {
            throw new IllegalArgumentException("Fant ikke Ident i Header ved sjekk av tilgang til Ident");
        } else {
            metrikk.tellHendelse("accessToIdent");

            return brukertilgangConsumer.hasAccessToAnsatt(oppslaattIdent);
        }
    }
}

File: src/main/kotlin/no/nav/syfo/brukertilgang/BrukerTilgangController.kt
package no.nav.syfo.brukertilgang

import no.nav.security.token.support.core.api.ProtectedWithClaims
import no.nav.security.token.support.core.context.TokenValidationContextHolder
import no.nav.syfo.metric.Metrikk
import no.nav.syfo.oidc.OIDCIssuer
import no.nav.syfo.oidc.OIDCUtil
import no.nav.syfo.pdl.PdlConsumer
import no.nav.syfo.service.BrukertilgangService
import no.nav.syfo.util.NAV_PERSONIDENT_HEADER
import no.nav.syfo.util.NAV_PERSONIDENT_HEADER_DEPRECATED
import org.slf4j.LoggerFactory
import org.springframework.util.MultiValueMap
import org.springframework.util.StringUtils
import org.springframework.web.bind.annotation.*
import javax.inject.Inject
import javax.ws.rs.ForbiddenException


@RestController
@ProtectedWithClaims(issuer = OIDCIssuer.EKSTERN)
@RequestMapping(value = ["/api/tilgang"])
class BrukerTilgangController @Inject constructor(
    private val contextHolder: TokenValidationContextHolder,
    private val brukertilgangConsumer: BrukertilgangConsumer,
    private val brukertilgangService: BrukertilgangService,
    private val pdlConsumer: PdlConsumer,
    private val metrikk: Metrikk
) {
    @GetMapping
    fun harTilgang(@RequestParam(value = "fnr", required = false) oppslaattFnr: String?): RSTilgang {
        val innloggetIdent = OIDCUtil.getSubjectEksternMedThrows(contextHolder)
        val oppslaattIdent = if (StringUtils.isEmpty(oppslaattFnr)) innloggetIdent else oppslaattFnr
        if (!brukertilgangService.tilgangTilOppslattIdent(innloggetIdent, oppslaattIdent)) {
            LOG.error("Ikke tilgang: Bruker spør om noen andre enn seg selv eller egne ansatte")
            throw ForbiddenException()
        }
        metrikk.tellHendelse("sjekk_brukertilgang")
        val isKode6Or7 = pdlConsumer.isKode6Or7(oppslaattIdent!!)
        return if (isKode6Or7) {
            RSTilgang(
                false,
                IKKE_TILGANG_GRUNN_DISKRESJONSMERKET
            )
        } else {
            RSTilgang(true)
        }
    }

    @GetMapping(path = ["/ansatt"])
    @ResponseBody
    fun accessToAnsatt(@RequestHeader headers: MultiValueMap<String, String>): BrukerTilgang {
        val oppslaattIdent = headers.getFirst(NAV_PERSONIDENT_HEADER_DEPRECATED.toLowerCase())
            ?: headers.getFirst(NAV_PERSONIDENT_HEADER)
        if (oppslaattIdent.isNullOrEmpty()) {
            throw IllegalArgumentException("Fant ikke Ident i Header ved sjekk av tilgang til Ident")
        } else {
            metrikk.tellHendelse("accessToIdent")
            return BrukerTilgang(brukertilgangConsumer.hasAccessToAnsatt(oppslaattIdent))
        }
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(BrukerTilgangController::class.java)
        const val IKKE_TILGANG_GRUNN_DISKRESJONSMERKET = "kode6-7"
    }

}
